### PR TITLE
Dynamically Add and Call Models

### PR DESCRIPTION
**Pull request by montasaurus**
### Code of Conduct

- [ ] I agree to follow this project's [Code of Conduct](https://github.com/OpenRouterTeam/openrouter-runner/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] I agree to license this contribution under the MIT LICENSE
- [ ] I checked the [current PR](https://github.com/OpenRouterTeam/openrouter-runner/pulls) for duplication.

### CODE REVIEW

File: modal/runner/containers/lorean_7b.py [hunk 1]
```diff
@@ -26,8 +26,8 @@
 
 @stub.cls(
     volumes={
-        str(loras_path): stub.loras_volume,
-        str(models_path): stub.models_volume,
+        loras_path: loras_volume,
+        models_path: models_volume,
```
Comment (louisgv) [inline modal/runner/containers/lorean_7b.py:30]: Ooh this actually work now? When I tried it back in August, the volume didn't register when spawning the GPU container.

File: modal/runner/engines/vllm.py [hunk 2]
```diff
@@ -61,7 +62,9 @@ def __init__(self, params: VllmParams):
     #     return engine_model_config.max_model_len
 
     @method()
-    async def generate(self, payload: Payload, params):
+    async def generate(self, payload: CompletionPayload, params):
+        models_volume.reload()
```
Comment (louisgv) [inline modal/runner/engines/vllm.py]: Ooh why is this needed?

File: scripts/shared.ts [hunk 3]
```diff
@@ -18,31 +30,29 @@ export async function completion(
...
     params: { max_tokens, stop },
     stream
   };
 
-  const p = await fetch(url, {
+  if (container) {
+    bodyPayload['runner'] = { container };
+  }
+
+  const p = await fetch(getApiUrl(''), {
```
Comment (louisgv) [inline scripts/shared.ts]: nit: maybe we should move the default uri into the params :-?

File: scripts/shared.ts [hunk 4]
```diff
@@ -54,6 +64,45 @@ export async function completion(
...
+  };
+
+  const response = await fetch(getApiUrl('/models'), {
+    method: 'POST',
+    headers: getAuthHeaders(),
+    body: JSON.stringify(payload)
+  });
+
+  if (!response.ok) {
+    throw new Error('Failed to post model: ' + response.status);
```
Comment (louisgv) [inline scripts/shared.ts:88]: We should grab the res.text() and include it in the error imo

File: scripts/shared.ts [hunk 5]
```diff
@@ -54,6 +64,45 @@ export async function completion(
...
+  }
+
+  return await response.json();
+}
+
+export async function awaitJob(jobId: string, timeoutMs: number) {
```
Comment (louisgv) [inline scripts/shared.ts]: nit: when calling this by itself, we would have double await which is weird to read:

```
await awaitJob(id)
```

Maybe it's better named `waitForJob`

File: scripts/shared.ts [hunk 6]
```diff
@@ -54,6 +64,45 @@ export async function completion(
...
+  const end = start + timeoutMs;
+  while (Date.now() < end) {
+    const statusResponse = await fetch(getApiUrl(`/jobs/${jobId}`), {
+      headers: getAuthHeaders()
+    });
+    if (statusResponse.status === 200) {
+      console.log('Job completed successfully');
+      break;
+    }
+    if (statusResponse.status != 202) {
```
Comment (louisgv) [inline scripts/shared.ts]: nit: use strict equal here

File: scripts/shared.ts [hunk 7]
```diff
@@ -54,6 +64,45 @@ export async function completion(
...
+      throw new Error('Failed to process job: ' + statusResponse.status);
+    }
+
+    console.log('Job still in progress...');
+
+    await new Promise((resolve) => setTimeout(resolve, 5000));
```
Comment (louisgv) [inline scripts/shared.ts]: nit: I wonder if we should pull the pooling interval out into a params

File: scripts/test-add-model.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+async function main(model?: string) {
+  const modelName = model || process.env.MODEL!;
+  console.log(`Test adding model ${modelName}`);
+  const body = await enqueueAddModel(modelName);
+  console.log('Successfully queued model to add', body);
+
+  // fetch job success or failure for up to the timeout of 1 hour
+  const timeoutMs = 60 * 60 * 1000;
+  await awaitJob(body.job_id, timeoutMs);
```
Comment (louisgv) [inline scripts/test-add-model.ts]: nit: The double await makes it weird to read

File: scripts/shared.ts [hunk 9]
```diff
@@ -18,31 +30,29 @@ export async function completion(
...
-  if (!url || !apiKey) {
-    throw new Error('Missing url or key');
-  }
-
   if (!quiet) {
     console.info(`Calling ${url} with model ${model}, stream: ${stream}`);
   }
 
-  const bodyPayload: Record<string, unknown> = {
+  let bodyPayload: Record<string, unknown> = {
```
Comment (louisgv) [inline scripts/shared.ts]: nit: I think we can keep const here since js objects are references, not actual struct. Thus inner mutation can be done to a const reference. I think it's better to leverage as much immutability as we can.

**Review by louisgv — COMMENTED** [review]
Commented

File: modal/runner/containers/lorean_7b.py (hunk 1 above)
Comment (montasaurus) [reply modal/runner/containers/lorean_7b.py:30]: The other containers have been working! Not sure about this Lorean one specifically, it seems unused?

File: modal/runner/engines/vllm.py (hunk 2 above)
Comment (montasaurus) [reply modal/runner/engines/vllm.py]: I'm getting intermittent file not found errors from the runners after adding new models. Suspect that it's due to this: https://modal.com/docs/guide/volumes#volume-commits-and-reloads

Going to look at doing this more efficiently, not sure if adding reloads on each generate call will cause issues under load

File: scripts/shared.ts (hunk 9 above)
Comment (montasaurus) [reply scripts/shared.ts]: Good point!

File: scripts/shared.ts (hunk 3 above)
Comment (montasaurus) [reply scripts/shared.ts]: I left the completion endpoint at the root so it's not a breaking change, but seems like the completion endpoint should be under e.g. `/completion` instead--what do you think?

If it had its own endpoint instead of the root, default argument would probably be more confusing?

File: scripts/shared.ts (hunk 5 above)
Comment (montasaurus) [reply scripts/shared.ts]: idk, this kinda makes my eyes bleed even more
```
await waitForJob(id)
```

File: scripts/shared.ts (hunk 6 above)
Comment (montasaurus) [reply scripts/shared.ts]: What if we return a status code of `202.0000` for added precision? 😆

File: scripts/shared.ts (hunk 7 above)
Comment (montasaurus) [reply scripts/shared.ts]: Make sense

File: modal/runner/shared/download.py [hunk 10]
```diff
@@ -15,47 +16,42 @@
...
+        patterns.append("*.bin")
+    # TODO: Use these patterns?
+
+    # Clean doesn't remove the cache, so using `local_files_only` here returns the cache even when the local dir is empty.
+    print(f"Checking for {model_name}")
+    snapshot_download(
+        model_name,
+        local_dir=model_path,
+        cache_dir=cache_path,
+        local_dir_use_symlinks=False,
```
Comment (louisgv) [inline modal/runner/shared/download.py]: I think we should use sym_links here to download the file once to cache only :-?... - and maybe we should fix the clean method to purge cache instead?

File: scripts/shared.ts (hunk 3 above)
Comment (louisgv) [reply scripts/shared.ts]: That sounds more consistent long term to me. I think we should test on dev to see if it would replace the old deployment URL. If it does, we might need to devise a migration strategy for or-web project.

Comment (montasaurus) [reply scripts/shared.ts]: Verified locally and it keeps the same URL thankfully
